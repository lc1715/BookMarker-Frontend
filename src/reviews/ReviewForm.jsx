import { useState, useContext } from 'react';
import ShowAlert from "../common/ShowAlert";
import BookMarkerApi from "../api/api";
import UserContext from "../auth/UserContext";
//Material UI
import Container from '@mui/material/Container';
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import FormControl from '@mui/material/FormControl';
import TextField from '@mui/material/TextField';


/**ReviewForm component
 * - Shows review form and allows users to write a review on a book
 * - Component using ReviewForm: BookDetail
 */
function ReviewForm({ book, addBook, setReviewChange, setOpenReviewForm, setBookStatus, setScrollToReview }) {
    const [formData, setFormData] = useState({
        comment: ''
    });
    const [formErrors, setFormErrors] = useState([]);
    const { currentUser, hasSavedBook } = useContext(UserContext);

    /**Handles form submission 
     * - If user writes a review on a book but the book has not been saved, 
     *  then automatically save book for user under 'Read' status 
     * - Set the openReviewForm state and scrollToReview state to scroll to 
     *  the user's review the very first time it is rendered
     */
    async function handleSubmit(evt) {
        evt.preventDefault();
        if (!hasSavedBook(book.volumeId)) {
            await addBook(book, true);
            setBookStatus(true);
        }

        try {
            await BookMarkerApi.addReview(book.volumeId, currentUser.username, { ...formData });
            setReviewChange(true);
            setOpenReviewForm(false);
            setScrollToReview(true);
        } catch (err) {
            setFormErrors(err);
            return;
        }
    };

    /**Handles changes in form data */
    function handleChange(evt) {
        const { name, value } = evt.target;

        setFormData((data) => ({
            ...data, [name]: value
        }));
    };

    return (
        <Container maxWidth="xl" sx={{ display: 'flex' }}>
            <form onSubmit={handleSubmit}>
                <FormControl
                    variant="outlined"
                    sx={{
                        width: { xs: 320, sm: 400, md: 555, lg: 575 },
                        height: 200,
                        mb: 2
                    }}>
                    <TextField
                        name='comment'
                        value={formData.comment}
                        label="Add Review"
                        onChange={handleChange}
                        rows={7}
                        multiline
                    />
                </FormControl>

                <Button variant="contained"
                    type='submit'
                    sx={{ fontSize: { lg: '13px' } }}
                >
                    Add Review
                </Button>

                <Button variant="contained"
                    size="medium"
                    sx={{ ml: 1, fontSize: { lg: '13px' } }}
                    type='button'
                    onClick={() => (setOpenReviewForm(false))}
                >
                    Cancel
                </Button>

                <Box sx={{ mt: 2, display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
                    {formErrors.length ? <ShowAlert type='error' messages={formErrors} /> : null}
                </Box>
            </form >
        </Container >
    )
};

export default ReviewForm;
