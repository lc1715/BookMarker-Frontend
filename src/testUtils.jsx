import UserContext from "./auth/UserContext";

const demoUser = {
    id: 1,
    username: 'testUser',
    email: 'test@email.com',
    volume_ids: ['e3', 't5'],
}

const UserProvider =
    ({ children, currentUser = demoUser, hasSavedBook = () => (true) }) => (
        <UserContext.Provider value={{ currentUser, hasSavedBook }}>
            {children}
        </UserContext.Provider>
    );

export { UserProvider };
