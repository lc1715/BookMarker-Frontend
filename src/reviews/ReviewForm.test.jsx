import { it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from "../testUtils";
import ReviewForm from './ReviewForm';

it('renders the ReviewForm component', () => {
    render(
        <MemoryRouter>
            <UserProvider>
                <ReviewForm />
            </UserProvider>
        </MemoryRouter>
    )
})

it('matches with snapshot', async () => {
    const { asFragment } = render(
        <MemoryRouter>
            <UserProvider>
                <ReviewForm />
            </UserProvider>
        </MemoryRouter>
    )

    expect(asFragment()).toMatchSnapshot();
})

it('has correct review form text', async () => {
    render(
        <MemoryRouter>
            <UserProvider>
                <ReviewForm />
            </UserProvider>
        </MemoryRouter>
    )

    expect(screen.getByLabelText('Add Review')).toBeInTheDocument();
});
