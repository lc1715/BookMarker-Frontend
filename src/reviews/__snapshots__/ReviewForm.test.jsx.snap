// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches with snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthXl css-pzt2s-MuiContainer-root"
  >
    <form>
      <div
        class="MuiFormControl-root css-1hp7dvp-MuiFormControl-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
          >
            Add Review
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-xrmkj5-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <textarea
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-w4nesw-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              name="comment"
              rows="7"
              style="height: 0px; overflow: hidden;"
            />
            <textarea
              aria-hidden="true"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-w4nesw-MuiInputBase-input-MuiOutlinedInput-input"
              readonly=""
              style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
              tabindex="-1"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-81qg8w"
              >
                <span>
                  Add Review
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-15a7cck-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Add Review
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1yqegjj-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Cancel
      </button>
      <div
        class="MuiBox-root css-ny6if"
      />
    </form>
  </div>
</DocumentFragment>
`;
